File: src/lib/api/globalSearch.ts
import { apiGet } from '@/lib/api/client.ts'

export type GlobalSearchKind =
  | 'rd_bill'
  | 'channel_bill'
  | 'contract'
  | 'invoice'
  | 'partner'
  | 'bank_transaction'

export type GlobalSearchTarget = {
  action: 'bill360' | 'contract_detail' | 'invoice_detail' | 'partner_focus' | 'bank_detail'
  view: string
  entity_id: string
  bill_type?: 'rd' | 'channel' | null
  direction?: 'input' | 'output' | null
  focus_query?: string | null
}

export type GlobalSearchResult = {
  id: string
  kind: GlobalSearchKind
  title: string
  subtitle?: string | null
  meta?: string | null
  badge?: string | null
  amount?: number | null
  status?: string | null
  score: number
  matched_fields: string[]
  target: GlobalSearchTarget
}

export type GlobalSearchResponse = {
  query: string
  total: number
  results: GlobalSearchResult[]
  groups: Array<{ kind: GlobalSearchKind; count: number }>
}

export function globalSearch(query: string, limit = 30) {
  const params = new URLSearchParams({ q: query, limit: String(limit) })
  return apiGet<GlobalSearchResponse>(`/api/global-search?${params.toString()}`, { timeoutMs: 12_000 })
}

File: backend/app/schemas/global_search.py
"""全局业务搜索响应模型。"""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field

SearchKind = Literal[
    "rd_bill",
    "channel_bill",
    "contract",
    "invoice",
    "partner",
    "bank_transaction",
]

SearchAction = Literal[
    "bill360",
    "contract_detail",
    "invoice_detail",
    "partner_focus",
    "bank_detail",
]


class GlobalSearchTarget(BaseModel):
    action: SearchAction
    view: str
    entity_id: str
    bill_type: Literal["rd", "channel"] | None = None
    direction: Literal["input", "output"] | None = None
    focus_query: str | None = None


class GlobalSearchResult(BaseModel):
    id: str
    kind: SearchKind
    title: str
    subtitle: str | None = None
    meta: str | None = None
    badge: str | None = None
    amount: float | None = None
    status: str | None = None
    score: int = Field(ge=0, le=100)
    matched_fields: list[str] = Field(default_factory=list)
    target: GlobalSearchTarget


class GlobalSearchGroupCount(BaseModel):
    kind: SearchKind
    count: int = Field(ge=0)


class GlobalSearchResponse(BaseModel):
    query: str
    total: int = Field(ge=0)
    results: list[GlobalSearchResult]
    groups: list[GlobalSearchGroupCount]
